src/obstacle_avoidance/src/MyNode.cpp

```diff
@@ -0,0 +1,231 @@
+          double C = lidar_data.angle_increment;
+          // Cosine Rule
+          double dist_between_pts =
+              sqrt(pow(a, 2) + pow(b, 2) - 2 * a * b * cos(C));
+          if ((dist_between_pts > MIN_GOAL_WIDTH) &&
+              (dist_between_pts < MAX_GOAL_WIDTH)) {
+            if (i < (lidar_array_size / 2 - 10)) {
+              left_side = FALSE;
+            } else if (i > (lidar_array_size / 2 + 10)) {
+              left_side = TRUE;
+            }
+          }
+          x = 1;
+        } else {
+          x++;
+        }
+        i++;
+      }
+    }
+  }
+
+  return left_side;
+}
+
+double AvoidObstacleClass::angVelocityToAvoidCollision() {
+
+  double critical_distance = MIN_CRITICAL_DISTANCE;
+  double ang_velocity = 0;
+  avoidance move_to_avoid_obstacle;
+  // std::cerr << "angV 1st action" << std::endl;
+
+  for (int i = 0; i < sizeOfLidarArray(lidar_data); i++) {
+    if ((lidar_data.ranges[i] != NAN) && (lidar_data.ranges[i] > 0.03) &&
+        (lidar_data.ranges[i] < 5)) {
+      // check if lidar point is going to collide
+      // std::cerr << "angV 2nd action" << std::endl;
+      move_to_avoid_obstacle = moveToAvoidCollision(
+          lidar_data.ranges[i],
+          (lidar_data.angle_min + lidar_data.angle_increment * i));
+      // std::cerr << "angV 3rd action" << std::endl;
+      // if most immediate threat/closest point
+      if (move_to_avoid_obstacle != no_avoidance) {
+        if (lidar_data.ranges[i] < critical_distance) {
+          critical_distance = lidar_data.ranges[i];
+          if (move_to_avoid_obstacle == left) {
+            // turn with x speed where x is dependent on how close the point is
+            ang_velocity =
+                (MIN_CRITICAL_DISTANCE - critical_distance) * ANG_VEL_GAIN;
+            // std::cout << "turn left " << ang_velocity << std::endl;
+
+          } else {
+            ang_velocity =
+                -(MIN_CRITICAL_DISTANCE - critical_distance) * ANG_VEL_GAIN;
+            // std::cout << "turn right " << ang_velocity << std::endl;
+          }
+        }
+      } else {
+        // std::cout << "no avoidance" << std::endl;
+      }
+    }
+  }
+  std::cout << "ang vel " << ang_velocity << std::endl;
+  return ang_velocity;
+}
+
+void AvoidObstacleClass::lidarCallback(
+    const sensor_msgs::LaserScan::ConstPtr &laser_scan_message) {
+  // get the lidar data from the subscribed topic
+  lidar_data.angle_min = laser_scan_message->angle_min;
+  lidar_data.angle_max = laser_scan_message->angle_max;
+  lidar_data.angle_increment = laser_scan_message->angle_increment;
+  // std::cout << "lidar data " << lidar_data.angle_max << std::endl;
+  // std::cout << "lidar scan " << laser_scan_message->angle_max << std::endl;
+
+  int lidar_array_size = sizeOfLidarArray(lidar_data);
+
+  lidar_data.ranges.resize(lidar_array_size);
+
+  for (int i = 0; i < lidar_array_size; i++) {
+    lidar_data.ranges[i] = laser_scan_message->ranges[i]; // edit
+  }
```

Comment:
Ideally, we want to trigger all the logic from this callback. This means that from here we should call something like `createAndPublishTwistMessage(*laser_scan_message)` which would determine the twist message needed based on the laserscan and publish it. This ensures that we are only re-computing the angular and linear velocities once per lidar scan. This pattern is not super important here, but it's a good habit to get into because it:
a) makes code easier to follow
b) larger data structures that we perform computationally intensive tasks on (such as images), it can save a lot of computational time